### PR TITLE
Adds product query support for Category list block

### LINKED ISSUE #7329: Product Query: [Atomic Block] Convert Category List Block
Refactor the Category List block to support the new Product Query block.

Requirements:

- Must be client-side rendered in the editor.
- Must be server-side rendered (with hydration support) on the frontend.
- The block should be kept independent from the Query Loop block for use with future blocks - see the Single Product block as an example.

### PR DESCRIPTION
**Pull request by imanish003**
Adds Product Query support for the atomic category list block.

On the client side, when the Category list Block is used within the Product Query block, the markup will be rendered on the server side - no javascript related to the Category list block will be loaded.

Fixes woocommerce/woocommerce-blocks#7329 

### Testing
1. Checkout this branch and run `npm start` 
2. Be sure that you have latest version of the Gutenberg feature plugin installed and activated
3. Create a post/page and add the Product Query block.
4. Add `woocommerce/product-category-list` block to `Post Template` of product query block
    - Verify that product category block works as it was working before.  
6. Now save the post & Open the frontend side and confirm that:
    - No JS related to the Product Category list block is loaded (search in your network tab for `product-category-list-frontend.js`).
    - The UI of the block is identical to on the editor side.
    - That the settings are reflected correctly (eg: custom style).
7. For comparison, switch to trunk and add the All Products block to confirm that the aforementioned product-category-list-frontend.js is loaded.

* [X] Do not include in the Testing Notes 

### WooCommerce Visibility

* [ ] WooCommerce Core
* [ ] Feature plugin
* [X] Experimental

### CODE REVIEW

**Comment by imanish003** [comment]
Hi @gigitux @sunyatasattva 👋 
This is the first time I have written some PHP code. Therefore, I have created a draft PR instead of a normal PR. 
Can you please have a look and provide me with some feedback? 🙂

Also, I noticed that UI is looking a little different between Editor & Frontend:
| Editor | Frontend |
| - | - |
| image | image | 

Can you please help me with this? 🙂

File: src/BlockTypes/ProductCategoryList.php [hunk 1]
```diff
@@ -48,12 +48,57 @@ protected function get_block_type_supports() {
...
+		$output  = '';
+		$output .= '
+			<div class="wc-block-components-product-category-list">
+				' . __( 'Categories:', 'woo-gutenberg-products-block' )
+				. '<ul>';
+
+		foreach ( $product_categories_terms as $product_category_term ) {
+			$output .= '
+				<li class="category-list-item-' . $product_category_term->slug . '">
+					<a href="' . get_term_link( $product_category_term->term_id ) . '">' . $product_category_term->name . '</a>
```
Comment (gigitux) [inline src/BlockTypes/ProductCategoryList.php]: The problem with the spaces has been a little bit tricky. In the end, this is the fix:

```suggestion
					<a href="' . get_term_link( $product_category_term->term_id ) . '">' . $product_category_term->name . '</a></li>
```

I found different links that discuss this problem:
- https://stackoverflow.com/questions/5078239/how-to-remove-the-space-between-inline-inline-block-elements
- https://css-tricks.com/fighting-the-space-between-inline-block-elements/


image

File: src/BlockTypes/ProductCategoryList.php [hunk 2]
```diff
@@ -48,12 +48,57 @@ protected function get_block_type_supports() {
...
+				</li>
```
Comment (gigitux) [inline src/BlockTypes/ProductCategoryList.php]: Obviously, if you want to adopt the solution above, you have to remove this tag.

**Review by gigitux — CHANGES_REQUESTED** [review]
Thanks for your work! You did a great job! 💪 

I noticed that the custom style doesn't work on the frontend side.

image
*editor side*

image
*frontend side*

The `StyleAttributesUtils` class helps you generate the right class/CSS to apply to elements.
Also, I noticed we have the same problem when the block is used inside the `All Products` block. Can you check it and open an issue (or fix it)?

The `Categories` word could have a custom translation. Did you check on this block to see if issue #7511 is replicable?

File: src/BlockTypes/ProductCategoryList.php (hunk 1 above)
Comment (imanish003) [reply src/BlockTypes/ProductCategoryList.php]: Ah, Got it. Thanks for sharing the link. The info mentioned [here](https://css-tricks.com/fighting-the-space-between-inline-block-elements/#aa-remove-the-spaces) is useful. 🙌🏻

File: src/BlockTypes/ProductCategoryList.php (hunk 2 above)
Comment (imanish003) [reply src/BlockTypes/ProductCategoryList.php]: Got it, Updating the highlighted lines to the following fixed it:
```php
			$output .= '
				<li class="category-list-item-' . $product_category_term->slug . '">
					<a href="' . get_term_link( $product_category_term->term_id ) . '">' . $product_category_term->name . '</a></li>
			';

```

**Comment by imanish003** [comment]
Hi @gigitux , Thanks for providing the solution for the spacing issue 🙏 

> Also, I noticed we have the same problem when the block is used inside the All Products block. Can you check it and open an issue (or fix it)?

Looks like this issue is also there on `trunk` branch; therefore, I have created a new issue for this: woocommerce/woocommerce#42526

**Comment by imanish003** [comment]
Hi @gigitux ,

> The Categories word could have a custom translation. Did you check on this block to see if issue https://github.com/woocommerce/woocommerce-blocks/issues/7511 is replicable?

This issue is not replicating for this PR. I believe this might be because I used [following line](https://github.com/woocommerce/woocommerce-blocks/blob/bac8cb2a90e55e78277a1907b3d4189805add186/src/BlockTypes/ProductCategoryList.php#L93):
```
. __( 'Categories:', 'woo-gutenberg-products-block' )
```

Just to be sure that this issue is not replicating, can you please try replicating this issue too?

**Comment by imanish003** [comment]
Hi @gigitux,
I have resolved the issue related to spacing in commit [d3b95f5](https://github.com/woocommerce/woocommerce-blocks/pull/7675/commits/d3b95f51d19c28a6f5265eb12d9f588d200bf386) and related to custom styling in commit [bac8cb2](https://github.com/woocommerce/woocommerce-blocks/pull/7675/commits/bac8cb2a90e55e78277a1907b3d4189805add186). 
Can you please review again? 🙏

**Comment by gigitux** [comment]
Thanks for your effort, Manish!

Regarding the style, there is a problem with the link color:

image

> Looks like this issue is also there on trunk branch; therefore, I have created a new issue for this: https://github.com/woocommerce/woocommerce/issues/42526

Thanks for opening a new issue!


> Just to be sure that this issue is not replicating, can you please try replicating this issue too?

I'm able to replicate this issue. I switched to the German language, and the block isn't translated:

image

The translation is available in German language: https://translate.wordpress.org/projects/wp-plugins/woo-gutenberg-products-block/dev/de/default/

**Comment by imanish003** [comment]
Hi @gigitux, Thanks for the thorough review 🙏🏻

> Regarding the style, there is a problem with the link color:

I have pushed a fix in commit [b046ec5](https://github.com/woocommerce/woocommerce-blocks/pull/7675/commits/b046ec52149bed18c803b43fd4c58ba55f2d0dbb) after our internal discussion in slack: p1669028757623899-slack-C02UBB1EPEF

> I'm able to replicate this issue. I switched to the German language, and the block isn't translated:

As discussed in our huddle call, we already have [following issue](https://github.com/woocommerce/woocommerce-blocks/issues/7511) created in past which we will handle separately.

**Review by gigitux — CHANGES_REQUESTED** [review]
Thanks for addressing the feedback! 🙇 

There is an error during the save when the `Product Categories List` block doesn't have a custom class.

image

To avoid this error, we should check on [this line](https://github.com/woocommerce/woocommerce-blocks/pull/7675/files#diff-cba89eae151ce52689cc5094ef24b697acea642796062009a14924d9ca0343c3R91) that `$attributes['className']` is defined.

**Comment by imanish003** [comment]
Hi @gigitux,
Nice catch. Thank you 🙏🏻 
I fixed the same issue in [sale badge PR](https://github.com/woocommerce/woocommerce-blocks/pull/7709) too 🙂

I have fixed the issue in commit [f03842e](https://github.com/woocommerce/woocommerce-blocks/pull/7675/commits/f03842e7502346f11b558bee155ca26e92f608e9). Please have a look again 🙏🏻

**Review by gigitux — APPROVED** [review]
Thanks for working on this! Great job! LGTM! 🚢

**Pull request merged**